README.md
# Winmin Viewer

Winmin Viewer is an application that uses the [SPICE protocol](https://www.spice-space.org/) to display applications from [libvirt](https://libvirt.org/) virtual machines.

# Build Dependencies

- libspice-client-gtk-3.0-dev
- libspice-client-glib-2.0-dev
- libgtk-3-dev
- meson

# Building

Build and install with:

```
$ meson build
$ ninja -C build install
```
# Runtime Dependencies

- spice-client-gtk

# Running

In order for this program to access the spice socket of the guest VM, the user must be part of the `kvm` group. This can be added using the following command.
```
$ sudo usermod -aG $USER kvm
```

You may need to logout or reboot in order for group changes to take effect.

### Usage
```
winmin-viewer name spice_socket
```
### Example
```
winmin-viewer "Word" spice+unix:///var/lib/libvirt/qemu/domain-1-winmin-msoffice/spice.sock
```
# TODO
- Format commandline arguments better see: [Parsing command-line arguments in C?](https://stackoverflow.com/questions/9642732/parsing-command-line-arguments-in-c)
- Add fullscreen option
- Add direct binding to libvirt virtual machines using the libvirt API
- Add USB passthrough options

debian/changelog
winmin-viewer (0.1-1) unstable; urgency=low

  * Initial release

 -- Victor Fuentes <hyruleterminatriforce@gmail.com>  Mon, 01 Jun 2020 00:17:52 +0100



winmin.c
#include <gtk/gtk.h>
#include <spice-client-gtk-3.0/spice-client-gtk.h>
#include <stdio.h>
#include <glib-object.h>
#include <spice-client.h>
#include <signal.h> 

SpiceSession * session;
char** name;
char** sock;
int id;
GtkWidget * window;
SpiceChannel * mainchannel;

gboolean resize_done (gpointer data)
{
  guint *gid = data;
  *gid = 0;

  resize();
  return FALSE;
}


gboolean on_configure_event (GdkEvent *event, gpointer data)
{
  static guint gid = 0;
  if (gid)
    g_source_remove (gid);
  gid = g_timeout_add (250, resize_done, &gid);
  return FALSE;
  
}

void resize() {
  printf("Resize!\n");
  int width, height;
  gtk_window_get_size(window, &width, &height);

  spice_main_update_display_enabled(mainchannel,0,TRUE,TRUE);
  spice_main_channel_update_display(mainchannel, 0, 0, 0, width, height, TRUE);
  printf("Send!\n");
  spice_main_send_monitor_config(mainchannel);

}




static void
activate (GtkApplication* app,
          gpointer        user_data)
{

  window = gtk_application_window_new (app);

  session = spice_session_new();

  g_object_set(session,"uri",sock);
     //"spice+unix:///var/lib/libvirt/qemu/domain-1-winmin-base/spice.sock");

  printf("test\n");
  char** test;
  g_object_get(session,"uri",&test);
  printf("test: %s\n",test);

  mainchannel =  spice_channel_new(session,1,id); // SPICE_CHANNEL(session);
  
  /*if (SPICE_IS_CHANNEL(mainchannel)) {
    printf("Is spice channel\n");
  }
  else {
    printf("Not spice channel\n");
    exit(1);
  }*/

  spice_session_connect(session);


  SpiceDisplay * display = spice_display_new(session, id);

  GtkWidget * displaywidget = GTK_WIDGET(display);
  gtk_container_add(GTK_CONTAINER(window),displaywidget);
    
  gtk_window_set_title (GTK_WINDOW(window), name);
  gtk_window_set_default_size (GTK_WINDOW (window), 900, 800);
  gtk_window_set_position(GTK_WINDOW(window), GTK_WIN_POS_CENTER);
  gtk_widget_show_all (window);

  resize();
  g_signal_connect(G_OBJECT(window), "configure-event",
        G_CALLBACK(on_configure_event), NULL);
}

int
main (int    argc,
      char **argv)
{
  GtkApplication *app;
  int status;
  name = argv[1];
  sock = argv[2];
  id = 0; //(int)strtol(argv[3], NULL, 10);
  app = gtk_application_new(NULL, G_APPLICATION_FLAGS_NONE); //NULL to allow multiple apps to run it at once
  
  g_signal_connect (app, "activate", G_CALLBACK (activate), NULL);

  status = g_application_run (G_APPLICATION (app), NULL, NULL);
  g_object_unref (app);

  return status;
}


winmin-viewer.c
#include <gtk/gtk.h>
#include <spice-client-gtk-3.0/spice-client-gtk.h>
#include <stdio.h>
#include <glib-object.h>
#include <spice-client.h>
#include <signal.h> 

SpiceSession * session;
char** name;
char** sock;
int id;
GtkWidget * window;
SpiceChannel * mainchannel;

gboolean resize_done (gpointer data)
{
  guint *gid = data;
  *gid = 0;

  resize();
  return FALSE;
}


gboolean on_configure_event (GdkEvent *event, gpointer data)
{
  static guint gid = 0;
  if (gid)
    g_source_remove (gid);
  gid = g_timeout_add (250, resize_done, &gid);
  return FALSE;
  
}

void resize() {
  printf("Resize!\n");
  int width, height;
  gtk_window_get_size(window, &width, &height);

  spice_main_update_display_enabled(mainchannel,0,TRUE,TRUE);
  spice_main_channel_update_display(mainchannel, 0, 0, 0, width, height, TRUE);
  printf("Send!\n");
  spice_main_send_monitor_config(mainchannel);

}




static void
activate (GtkApplication* app,
          gpointer        user_data)
{

  window = gtk_application_window_new (app);

  session = spice_session_new();

  g_object_set(session,"uri",sock);
     //"spice+unix:///var/lib/libvirt/qemu/domain-1-winmin-base/spice.sock");

  mainchannel =  spice_channel_new(session,1,id); // SPICE_CHANNEL(session);
  
  /*if (SPICE_IS_CHANNEL(mainchannel)) {
    printf("Is spice channel\n");
  }
  else {
    printf("Not spice channel\n");
    exit(1);
  }*/

  spice_session_connect(session);


  SpiceDisplay * display = spice_display_new(session, id);

  GtkWidget * displaywidget = GTK_WIDGET(display);
  gtk_container_add(GTK_CONTAINER(window),displaywidget);
    
  gtk_window_set_title (GTK_WINDOW(window), name);
  gtk_window_set_default_size (GTK_WINDOW (window), 900, 800);
  gtk_window_set_position(GTK_WINDOW(window), GTK_WIN_POS_CENTER);
  gtk_widget_show_all (window);

  resize();
  g_signal_connect(G_OBJECT(window), "configure-event",
        G_CALLBACK(on_configure_event), NULL);
}

int
main (int    argc,
      char **argv)
{
  GtkApplication *app;
  int status;
  name = argv[1];
  sock = argv[2];
  id = 0; //(int)strtol(argv[3], NULL, 10);
  app = gtk_application_new(NULL, G_APPLICATION_FLAGS_NONE); //NULL to allow multiple apps to run it at once
  
  g_signal_connect (app, "activate", G_CALLBACK (activate), NULL);

  status = g_application_run (G_APPLICATION (app), NULL, NULL);
  g_object_unref (app);

  return status;
}


winmin-multiarray.c
#include <gtk/gtk.h>
#include <spice-client-gtk-3.0/spice-client-gtk.h>
#include <stdio.h>
#include <glib-object.h>
#include <spice-client.h>
#include <signal.h> 

SpiceSession * session;
char** name;
char** sock;
int id;
int windowcount;
GtkWidget * window[2];
SpiceChannel * mainchannel;

gboolean resize_done (gpointer data)
{
  guint *gid = data;
  *gid = 0;

  resize();
  
  return FALSE;
}


gboolean on_configure_event (GdkEvent *event, gpointer data)
{
  static guint gid = 0;
  if (gid)
    g_source_remove (gid);
  gid = g_timeout_add (250, resize_done, &gid);
  return FALSE;
  
}

void resize() {
  printf("Resize!\n");

  for (int i = 0; i < windowcount; ++i) {
    int width, height;
    gtk_window_get_size(window[i], &width, &height);
    spice_main_set_display_enabled(mainchannel,i,TRUE);
    spice_main_set_display(mainchannel,i, 0, 0, width, height);

  }
  //spice_main_channel_update_display(mainchannel, 1, 0, 0, 500, 500, TRUE);
  printf("Send!\n");
  spice_main_send_monitor_config(mainchannel);
  //spice_main_set_display_enabled(mainchannel,-1,TRUE);


}


void setup_spice() {
  session = spice_session_new();
  g_object_set(session,"uri",sock);
  printf("test\n");
  char** test;
  g_object_get(session,"uri",&test);
  printf("test: %s\n",test);

  mainchannel =  spice_channel_new(session,1,0);
  spice_session_connect(session);
  if (SPICE_IS_MAIN_CHANNEL(mainchannel)) {
    printf("Good\n");
  }
  else {
    printf("Bad\n");
  }
}


static void
activate (GtkApplication* app,
          gpointer        user_data)
{

  window[windowcount] = gtk_application_window_new (app);

  SpiceDisplay * display = spice_display_new(session, windowcount);

  GtkWidget * displaywidget = GTK_WIDGET(display);
  gtk_container_add(GTK_CONTAINER(window[windowcount]),displaywidget);
    
  gtk_window_set_title (GTK_WINDOW(window[windowcount]), name);
  gtk_window_set_default_size (GTK_WINDOW (window[windowcount]), 900, 800);
  gtk_window_set_position(GTK_WINDOW(window[windowcount]), GTK_WIN_POS_CENTER);
  gtk_widget_show_all (window[windowcount]);

  resize();
  g_signal_connect(G_OBJECT(window[windowcount]), "configure-event",
        G_CALLBACK(on_configure_event), NULL);
  windowcount += 1;
}

int
main (int    argc,
      char **argv)
{
  GtkApplication *app;
  int status;
  name = argv[1];
  sock = argv[2];
  id = (int)strtol(argv[3], NULL, 10);
  app = gtk_application_new(NULL, G_APPLICATION_FLAGS_NONE);
  
  

  setup_spice();
  windowcount = 0;
  g_signal_connect (app, "activate", G_CALLBACK (activate), NULL);
  g_signal_connect (app, "activate", G_CALLBACK (activate), NULL);

  status = g_application_run (G_APPLICATION (app), NULL, NULL);

  g_object_unref (app);

  return status;
}




/*static void
activateaux (GtkApplication* app,
          gpointer        user_data)
{

  window[1] = gtk_application_window_new (app);

  SpiceDisplay * display2 = spice_display_new(session,1);

  GtkWidget * displaywidget = GTK_WIDGET(display2);
  gtk_container_add(GTK_CONTAINER(window[1]),displaywidget);
    
  gtk_window_set_title (GTK_WINDOW(window[1]), name);
  gtk_window_set_default_size (GTK_WINDOW (window[1]), 900, 800);
  gtk_window_set_position(GTK_WINDOW(window[1]), GTK_WIN_POS_CENTER);
  gtk_widget_show_all (window[1]);

  resize();
  g_signal_connect(G_OBJECT(window[1]), "configure-event",
        G_CALLBACK(on_configure_event), NULL);
}
*/


winmin-viewer-multimonitor.c
//WIP for winmin-viewer multimonitor support
#include <gtk/gtk.h>
#include <spice-client-gtk-3.0/spice-client-gtk.h>
#include <stdio.h>
#include <glib-object.h>
#include <spice-client.h>
#include <signal.h> 

SpiceSession * session;
char** name;
char** sock;
int id;
int windowcount;
GtkWidget * window[2];
SpiceChannel * mainchannel;

gboolean resize_done (gpointer data)
{
  guint *gid = data;
  *gid = 0;

  resize();
  
  return FALSE;
}


gboolean on_configure_event (GdkEvent *event, gpointer data)
{
  static guint gid = 0;
  if (gid)
    g_source_remove (gid);
  gid = g_timeout_add (250, resize_done, &gid);
  return FALSE;
  
}

void resize() {
  printf("Resize!\n");

  for (int i = 0; i < windowcount; ++i) {
    int width, height;
    gtk_window_get_size(window[i], &width, &height);
    spice_main_set_display_enabled(mainchannel,i,TRUE);
    spice_main_set_display(mainchannel,i, 0, 0, width, height);

  }
  //spice_main_channel_update_display(mainchannel, 1, 0, 0, 500, 500, TRUE);
  printf("Send!\n");
  spice_main_send_monitor_config(mainchannel);
  //spice_main_set_display_enabled(mainchannel,-1,TRUE);


}


void setup_spice() {
  session = spice_session_new();
  g_object_set(session,"uri",sock);
  printf("test\n");
  char** test;
  g_object_get(session,"uri",&test);
  printf("test: %s\n",test);

  mainchannel =  spice_channel_new(session,1,0);
  spice_session_connect(session);
  if (SPICE_IS_MAIN_CHANNEL(mainchannel)) {
    printf("Good\n");
  }
  else {
    printf("Bad\n");
  }
}


static void
activate (GtkApplication* app,
          gpointer        user_data)
{

  window[windowcount] = gtk_application_window_new (app);

  SpiceDisplay * display = spice_display_new(session, windowcount);

  GtkWidget * displaywidget = GTK_WIDGET(display);
  gtk_container_add(GTK_CONTAINER(window[windowcount]),displaywidget);
    
  gtk_window_set_title (GTK_WINDOW(window[windowcount]), name);
  gtk_window_set_default_size (GTK_WINDOW (window[windowcount]), 900, 800);
  gtk_window_set_position(GTK_WINDOW(window[windowcount]), GTK_WIN_POS_CENTER);
  gtk_widget_show_all (window[windowcount]);

  resize();
  g_signal_connect(G_OBJECT(window[windowcount]), "configure-event",
        G_CALLBACK(on_configure_event), NULL);
  windowcount += 1;
}

int
main (int    argc,
      char **argv)
{
  GtkApplication *app;
  int status;
  name = argv[1];
  sock = argv[2];
  id = (int)strtol(argv[3], NULL, 10);
  app = gtk_application_new(NULL, G_APPLICATION_FLAGS_NONE);
  
  

  setup_spice();
  windowcount = 0;
  g_signal_connect (app, "activate", G_CALLBACK (activate), NULL);
  g_signal_connect (app, "activate", G_CALLBACK (activate), NULL);

  status = g_application_run (G_APPLICATION (app), NULL, NULL);

  g_object_unref (app);

  return status;
}




/*static void
activateaux (GtkApplication* app,
          gpointer        user_data)
{

  window[1] = gtk_application_window_new (app);

  SpiceDisplay * display2 = spice_display_new(session,1);

  GtkWidget * displaywidget = GTK_WIDGET(display2);
  gtk_container_add(GTK_CONTAINER(window[1]),displaywidget);
    
  gtk_window_set_title (GTK_WINDOW(window[1]), name);
  gtk_window_set_default_size (GTK_WINDOW (window[1]), 900, 800);
  gtk_window_set_position(GTK_WINDOW(window[1]), GTK_WIN_POS_CENTER);
  gtk_widget_show_all (window[1]);

  resize();
  g_signal_connect(G_OBJECT(window[1]), "configure-event",
        G_CALLBACK(on_configure_event), NULL);
}
*/
